src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+    // be adding and removing things from diffferent sources, we convert
+    // everything to an `OsString` to make the lifetimes easier.
+    let tail: collections::VecDeque<ffi::OsString> = path.components()
+        .skip_while(is_head_part)
+        .map(|each| each.as_os_str().to_os_string())
+        .collect();
+
+    debug!("Head: {:?}, tail: {:?}", head, tail);
+    Ok((head, tail))
+}
+
+/// Returns the target of the symlink at `path`, if it exists and is one.
+///
+/// If `path` is a symlink and we can read the target, returns
+/// `Ok(Some(target))`.
+/// If `path` is not a symlink, returns `Ok(None)`.
+/// If `path` does not exist, it's still not a symlink and this function
+/// returns `Ok(None)`.
+/// Otherwise, returns the relevant error.
+fn read_link_if_exists<P: AsRef<path::Path>>(
+    path: P,
+) -> Result<Option<path::PathBuf>, Error> {
+    let path = path.as_ref();
+
+    // In theory, we could just call .read_link() directly and handle the error
+    // result, but:
+    //
+    // - on Windows, the "can't read this because it isn't a symlink"
+    //   error is returned as an io::ErrorKind::Other rather than
+    //   io::ErrorKind::InvalidInput.
+    // - on Windows, some not-smart filesystems (like VirtualBox's shared
+    //   folders) don't understand the API call, and just return "invalid
+    //   operation".
+    //
+    // Therefore, we'll check if the thing is a symlink before trying to read
+    // it.
+    path.symlink_metadata()
+        // If we successfully got metadata, check if it's a symlink.
+        .map(|metadata| metadata.file_type().is_symlink())
+        // If we failed to get metadata...
+        .or_else(|err| {
+            if err.kind() == io::ErrorKind::NotFound {
+                // ...and we failed because the path doesn't exist, by
+                // definition it cannot be a symlink.
+                debug!(
+                    "{:?} does not exist, it's not a symlink",
+                    path,
+                );
+                Ok(false)
+            } else {
+                // Any other error we can return as-is.
+                error!("Could not check {:?}", path);
+                Err(err)
+            }
+        })
+        // If we know for sure whether this is a symlink...
+        .and_then(|is_symlink| {
+            // ...and it *is* a symlink...
+            if is_symlink {
+                // ...let's read it to find the target.
+                debug!("{:?} exists and is a symlink", path);
+                Ok(Some(path.read_link()?))
+
+            // ...and it *isn't* a symlink...
+            } else {
+                // ...then obviously we don't have a target.
+                debug!("{:?} exists and isn't a symlink", path);
+                Ok(None)
+            }
+        })
+        .map_err(|err| Error::IoError {
+            err: err,
+            at: path.into(),
+        })
+}
+
+/// An absolute path that may or may not exist.
```

Comment:
Forgive me if I'm mistaken, but I believe a briefer description (since `PathAbs` already exists) is:

> Returns a `PathAbs` by directly following relative directory links while preserving symlinks.

IMO if this were to be included in `path_abs` it would be as a method `PathAbs::follow_relative(path: AsRef<Path>)` (am very open to better names).

Am I accurate in that and does that path forward (lol, pun) make sense?

The point I'm trying to make is that this _is_ a `PathAbs` (it follows the same rules) it is just constructed differently when there are relative paths involved in the constructor path.

Reply:
Yes, I expected/hoped that you would take `Absolute::new()`, rename it, and paste it into `PathAbs` as an alternative constructor. Or even (since this is my favourite path-absolutizing algorithm) as the default constructor, `new()`—but I leave that decsion up to you.

I think "relative" is a confusing term, since `foo/bar` is also a relative path and `PathAbs` and `Absolute` treat it the same way. The key difference between them is the treatment of `ParentDir` path components, so I'd probably name the constructor something like `PathAbs::with_parents_resolved()`.

Reply:
Having `PathAbs::with_parents_resolved<P: AsRef<Path>>(path: P)` sounds good to me :+1:

Reply:
I think I should start moving code into `src/abs.rs`, using the standard error enum, etc. rather than keeping this in a separate module. Does that sound reasonable?